cmd/tcplb/server.go

```diff
@@ -0,0 +1,223 @@
+package main
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"net"
+	"tcplb/lib/authn"
+	"tcplb/lib/authz"
+	"tcplb/lib/core"
+	"tcplb/lib/forwarder"
+	"tcplb/lib/limiter"
+	"tcplb/lib/slog"
+	"time"
+)
+
+const (
+	defaultAcceptErrorCooldownDuration = time.Second
+	defaultUpstreamNetwork             = "tcp"
+	defaultListenNetwork               = "tcp"
+	defaultListenAddress               = "0.0.0.0:4321"
+	defaultMaxConnectionsPerClient     = 10
+)
+
+var NotImplementedError = errors.New("not implemented")
+
+// TODO FIXME insecure
+var anonymousTestClientID = core.ClientID{Namespace: "test", Key: "anonymous"}
+
+type Config struct {
+	ListenNetwork           string
+	ListenAddress           string
+	Upstreams               []core.Upstream
+	MaxConnectionsPerClient int64
+}
+
+func (c *Config) Validate() error {
+	if len(c.Upstreams) == 0 {
+		return errors.New("server must be configured with 1 or more upstreams")
+	}
+	return nil
+}
+
+func makeClientReserverFromConfig(cfg *Config) (forwarder.ClientReserver, error) {
+	var reserver forwarder.ClientReserver
+	if cfg.MaxConnectionsPerClient > 0 {
+		reserver = limiter.NewUniformlyBoundedClientReserver(cfg.MaxConnectionsPerClient)
+	} else {
+		reserver = limiter.UnboundedClientReserver{}
+	}
+	return reserver, nil
+}
+
+func makeAuthorizerFromConfig(cfg *Config) (forwarder.Authorizer, error) {
+	// TODO FIXME begin placeholder demo authorization config
+	urGroup := authz.Group{Key: "ur"}
+	urUpstreamGroup := authz.UpstreamGroup{Key: "ur"}
+	authzCfg := authz.Config{
+		GroupsByClientID: map[core.ClientID][]authz.Group{
+			anonymousTestClientID: {urGroup},
+		},
+		UpstreamGroupsByGroup: map[authz.Group][]authz.UpstreamGroup{
+			urGroup: {urUpstreamGroup},
+		},
+		UpstreamsByUpstreamGroup: map[authz.UpstreamGroup]core.UpstreamSet{
+			urUpstreamGroup: core.NewUpstreamSet(cfg.Upstreams...),
+		},
+	}
+	// TODO FIXME end placeholder demo authorization config
+	return authz.NewStaticAuthorizer(authzCfg), nil
+}
+
+// IdiotDialer attempts to dial an arbitrary candidate and gives up if that fails.
+// This is a placeholder implementation with various issues:
+// - no timeout
+// - it doesn't attempt to balance load
+// - it doesn't try alternative upstreams if one attempt fails
+// - it doesn't learn anything
+type IdiotDialer struct {
```

Comment:
I get that this is likely throwaway, but let's be more careful with naming if we can.